File: twag/web/frontend/src/components/feed/articleVisuals.ts
import type { ArticleTopVisual, MediaItem } from "@/api/types";

export interface DataVisual {
  url: string;
  kind: string;
  isTop: boolean;
  keyTakeaway: string;
  whyImportant: string;
}

const DATA_KINDS = new Set(["chart", "table", "document", "screenshot"]);
const TEXTUAL_DATA_RE =
  /\b(chart|graph|table|capex|revenue|margin|growth|yoy|qoq|forecast|projection|run[-\s]?rate|backlog|roi|ebitda|eps)\b/;
const NUMERIC_DATA_RE =
  /(\$|\b\d+(\.\d+)?%|\b\d+(\.\d+)?\s?(b|m|bn|mn|trillion|billion|million)\b)/;
const NOISE_RE = /\b(meme|reaction image|shitpost|joke|selfie|portrait)\b/;

function buildTextBlob(item: MediaItem): string {
  return [
    item.short_description,
    item.prose_summary,
    item.prose_text,
    item.alt_text,
  ]
    .filter(Boolean)
    .join(" ")
    .toLowerCase();
}

function looksDataLikeText(text: string): boolean {
  if (!text) return false;
  if (NOISE_RE.test(text)) return false;
  return TEXTUAL_DATA_RE.test(text) || NUMERIC_DATA_RE.test(text);
}

function inferKind(item: MediaItem): string {
  const rawKind = (item.kind ?? "").toLowerCase();
  if (DATA_KINDS.has(rawKind)) return rawKind;
  if (item.chart?.description || item.chart?.insight || item.chart?.implication)
    return "chart";
  if (item.table?.columns?.length || item.table?.summary) return "table";
  if (looksDataLikeText(buildTextBlob(item))) return "chart";
  return rawKind;
}

function extractTakeaway(item: MediaItem, kind: string): string {
  if (kind === "chart") {
    return (
      item.chart?.insight ??
      item.chart?.implication ??
      item.chart?.description ??
      ""
    );
  }
  if (kind === "table") {
    return item.table?.summary ?? item.table?.description ?? "";
  }
  if (kind === "document" || kind === "screenshot") {
    return item.prose_summary ?? item.short_description ?? "";
  }
  return item.short_description ?? "";
}

function isRelevantVisual(item: MediaItem, kind: string): boolean {
  if (DATA_KINDS.has(kind)) return true;
  return looksDataLikeText(buildTextBlob(item));
}

export function buildArticleVisuals(
  topVisual: ArticleTopVisual | null,
  mediaItems: MediaItem[] | null,
  maxItems = 5,
): DataVisual[] {
  if (maxItems <= 0) return [];

  const visuals: DataVisual[] = [];
  const seen = new Set<string>();

  if (topVisual?.url) {
    const topKind = (topVisual.kind || "visual").toLowerCase();
    const topText =
      `${topVisual.key_takeaway || ""} ${topVisual.why_important || ""}`.toLowerCase();
    if (DATA_KINDS.has(topKind) || looksDataLikeText(topText)) {
      visuals.push({
        url: topVisual.url,
        kind: DATA_KINDS.has(topKind) ? topKind : "chart",
        isTop: true,
        keyTakeaway: topVisual.key_takeaway || "",
        whyImportant: topVisual.why_important || "",
      });
      seen.add(topVisual.url);
    }
  }

  const extras: Array<{ priority: number; visual: DataVisual }> = [];
  for (const item of mediaItems ?? []) {
    if (!item?.url || seen.has(item.url)) continue;
    const kind = inferKind(item);
    if (!isRelevantVisual(item, kind)) continue;
    const normalizedKind = DATA_KINDS.has(kind) ? kind : "chart";
    extras.push({
      priority:
        { chart: 0, table: 1, screenshot: 2, document: 3 }[normalizedKind] ?? 9,
      visual: {
        url: item.url,
        kind: normalizedKind,
        isTop: false,
        keyTakeaway: extractTakeaway(item, kind),
        whyImportant: "",
      },
    });
    seen.add(item.url);
  }

  extras.sort((a, b) => a.priority - b.priority);
  for (const extra of extras) {
    if (visuals.length >= maxItems) break;
    visuals.push(extra.visual);
  }

  return visuals;
}

File: twag/article_visuals.py
"""Helpers for selecting data-oriented visuals for article summaries."""

from __future__ import annotations

import re
from typing import Any

_DATA_KINDS = {"chart", "table", "document", "screenshot"}
_TEXTUAL_DATA_PATTERN = re.compile(
    r"\b(chart|graph|table|capex|revenue|margin|growth|yoy|qoq|forecast|projection|run[-\s]?rate|backlog|roi|ebitda|eps)\b"
)
_NUMERIC_DATA_PATTERN = re.compile(r"(\$|\b\d+(\.\d+)?%|\b\d+(\.\d+)?\s?(b|m|bn|mn|trillion|billion|million)\b)")
_NOISE_PATTERN = re.compile(r"\b(meme|reaction image|shitpost|joke|selfie|portrait)\b")


def _text_blob(*parts: Any) -> str:
    return " ".join(str(p).strip().lower() for p in parts if p is not None and str(p).strip())


def _looks_data_text(text: str) -> bool:
    if not text:
        return False
    if _NOISE_PATTERN.search(text):
        return False
    return bool(_TEXTUAL_DATA_PATTERN.search(text) or _NUMERIC_DATA_PATTERN.search(text))


def _infer_kind(item: dict[str, Any]) -> str:
    kind = str(item.get("kind") or "").strip().lower()
    if kind in _DATA_KINDS:
        return kind
    if isinstance(item.get("chart"), dict) and (
        item["chart"].get("description") or item["chart"].get("insight") or item["chart"].get("implication")
    ):
        return "chart"
    if isinstance(item.get("table"), dict) and (item["table"].get("columns") or item["table"].get("summary")):
        return "table"
    text_blob = _text_blob(
        item.get("short_description"),
        item.get("prose_summary"),
        item.get("prose_text"),
        item.get("alt_text"),
    )
    if _looks_data_text(text_blob):
        return "chart"
    return kind


def _extract_takeaway(item: dict[str, Any], kind: str) -> str:
    if kind == "chart":
        raw_chart = item.get("chart")
        chart = raw_chart if isinstance(raw_chart, dict) else {}
        return str(chart.get("insight") or chart.get("implication") or chart.get("description") or "").strip()
    if kind == "table":
        raw_table = item.get("table")
        table = raw_table if isinstance(raw_table, dict) else {}
        return str(table.get("summary") or table.get("description") or "").strip()
    if kind in {"document", "screenshot"}:
        return str(item.get("prose_summary") or item.get("short_description") or "").strip()
    return str(item.get("short_description") or "").strip()


def _is_relevant_visual(item: dict[str, Any], kind: str) -> bool:
    if kind in _DATA_KINDS:
        return True
    text_blob = _text_blob(
        item.get("short_description"),
        item.get("prose_summary"),
        item.get("prose_text"),
        item.get("alt_text"),
    )
    return _looks_data_text(text_blob)


def build_article_visuals(
    *,
    top_visual: dict[str, Any] | None,
    media_items: list[dict[str, Any]] | None,
    max_items: int = 4,
) -> list[dict[str, Any]]:
    """Build ordered article visuals: top visual first, then other data visuals."""
    if max_items <= 0:
        return []

    visuals: list[dict[str, Any]] = []
    seen_urls: set[str] = set()

    if isinstance(top_visual, dict):
        top_url = str(top_visual.get("url") or "").strip()
        if top_url:
            why_important = str(top_visual.get("why_important") or "").strip()
            key_takeaway = str(top_visual.get("key_takeaway") or "").strip()
            top_kind = str(top_visual.get("kind") or "").strip().lower() or "visual"
            top_text = _text_blob(why_important, key_takeaway)
            if top_kind in _DATA_KINDS or _looks_data_text(top_text):
                visuals.append(
                    {
                        "url": top_url,
                        "kind": top_kind if top_kind in _DATA_KINDS else "chart",
                        "is_top": True,
                        "why_important": why_important,
                        "key_takeaway": key_takeaway,
                    }
                )
                seen_urls.add(top_url)

    extras: list[tuple[int, dict[str, Any]]] = []
    if isinstance(media_items, list):
        for item in media_items:
            if not isinstance(item, dict):
                continue
            url = str(item.get("url") or "").strip()
            if not url or url in seen_urls:
                continue
            kind = _infer_kind(item)
            if not _is_relevant_visual(item, kind):
                continue
            takeaway = _extract_takeaway(item, kind)
            normalized_kind = kind if kind in _DATA_KINDS else "chart"
            priority = {"chart": 0, "table": 1, "screenshot": 2, "document": 3}.get(normalized_kind, 9)
            extras.append(
                (
                    priority,
                    {
                        "url": url,
                        "kind": normalized_kind,
                        "is_top": False,
                        "why_important": "",
                        "key_takeaway": takeaway,
                    },
                )
            )
            seen_urls.add(url)

    extras.sort(key=lambda pair: pair[0])
    for _, visual in extras:
        if len(visuals) >= max_items:
            break
        visuals.append(visual)

    return visuals
